### PR TITLE
feat: add CRC openshift lifecycle

### LINKED ISSUE #30: Ability to register Kubernetes provider
**Is your feature request related to a problem? Please describe.**
As we register container engines, we should be able to register kubernetes provider.

A kubernetes provider is a Kubernetes/OpenShift endpoint where we have access to the k8s API
Can be a local or remote endpoint

**Describe the solution you'd like**
Plug-ins can register kubernetes endpoint

Framework can list the endpoints

### PR DESCRIPTION
**Pull request by evidolob**
#29 
Fixes #30

----

Screenshot 2022-03-18 at 10 34 58

### CODE REVIEW

File: extensions/crc/src/daemon-commander.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
+const got = require('got');
+const os = require('os');
```
Comment (benoitf) [inline extensions/crc/src/daemon-commander.ts]: Hi, should probably follow 'import' styles there
+ copyright header

File: extensions/crc/src/daemon-commander.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    return JSON.parse(body);
+  }
+
+  async logs() {
+    const options = {
+      url: this.apiPath + '/logs',
+      method: 'GET',
+    };
+
+    const { body } = await got(options);
```
Comment (benoitf) [inline extensions/crc/src/daemon-commander.ts:57]: I don't see where this library is added in dependencies ?

File: packages/extension-api/src/extension-api.d.ts [hunk 3]
```diff
@@ -90,19 +90,25 @@ declare module '@tmpwip/extension-api' {
...
+    status(): ProviderStatus;
+    handleLifecycleChange(callback: (event: ProviderStatus) => void): Promise<void>;
   }
 
   export interface ContainerProvider {
     provideName(): string;
     provideConnection(): PromiseLike<string>;
   }
 
+  export interface ClusterProvider {
```
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]:
```suggestion
  export interface KubernetesProvider {
```

File: packages/extension-api/src/extension-api.d.ts [hunk 4]
```diff
@@ -112,9 +118,12 @@ declare module '@tmpwip/extension-api' {
...
   export namespace container {
     export function registerContainerProvider(provider: ContainerProvider): PromiseLike<Disposable>;
-    export function registerContainerProviderLifecycle(
-      providerLifecycle: ContainerProviderLifecycle,
-    ): PromiseLike<Disposable>;
+    export function registerContainerProviderLifecycle(providerLifecycle: ProviderLifecycle): PromiseLike<Disposable>;
+  }
+
+  export namespace kubernetes {
+    export function registerProvider(provider: ClusterProvider): PromiseLike<Disposable>;
```
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]:
```suggestion
    export function registerProvider(provider: KubernetesProvider): PromiseLike<Disposable>;
```

File: packages/extension-api/src/extension-api.d.ts [hunk 5]
```diff
@@ -112,9 +118,12 @@ declare module '@tmpwip/extension-api' {
...
+    export function registerLifecycle(clusterLifecycle: ClusterLifecycle): PromiseLike<Disposable>;
```
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts:126]: where is defined ClusterLifecycle ? shouldn't it be ProviderLifecycle ?

**Comment by slemeur** [comment]
Nice!

**Comment by rhatdan** [comment]
A couple of quick comments, would be better to call the drop down Kubernetes or OpenShift rather then CRC. 

I was hoping we would move to where the podman machine would just get enhanced with new layers for CRC rather then have a separate machine running.  But that does not effect the GUI.

**Comment by gbraad** [comment]
> Kubernetes or OpenShift rather than CRC

I agree, `crc` is just a tool to maintain the lifecycle of a VM wth a container runtime. Would it be possible to do inspection/guestimation if the cluster is kubernetes or OpenShift based? From the point of view when CRC is used it would now be OpenShift, but this is not an assumption that might hold in future. I prefer to be generic here.

File: extensions/crc/src/daemon-commander.ts (hunk 2 above)
Comment (gbraad) [reply extensions/crc/src/daemon-commander.ts:57]: `got` is the REST library used by `crc`. Perhaps the one used by `dockerode` could be used here.
The reason we use `got` has to do with named pipe usage on Windows.

**Comment by gbraad** [comment]
Can you clarify the states of the VM and the icon color? From the PR it looks as follows:

* busy 🟠
* stopped ⚫
* unknown 🔴
* started 🟢

Busy and started are similar to CRC. 
Just the states 'stopped' and 'unknown' have changed. 'Unknown' might mean 'errored' ?

For CRC we use 'unknown' also when no VM exists, as the state 'stopped' and 'unknown' only has an internal meaning, but for the user it is the same outcome. It needs to be started before the environment is available.

**Review by benoitf — APPROVED** [review]

File: extensions/crc/src/daemon-commander.ts (hunk 2 above)
Comment (benoitf) [reply extensions/crc/src/daemon-commander.ts:57]: here I was just saying that the library is used but the package was not added to the list of dependencies

for named pipe or socket you could even use pretty dumb code like
```
const net = require('net');
const conn = net.createConnection('\\\\.\\pipe\\pipe_name’);
```

**Comment by evidolob** [comment]
@gbraad You are right, this PR use those colours in the way that you describe.

**Comment by gbraad** [comment]
@evidolob there might be an identical issue as we have; the icons are not compatible with Apple's HiDPI use. Since there are several issues with the icons, it might be better to create a follow-up issue; for example, the menubar/tray icon itself is colored on macOS which is against their guidelines. WDYT @benoitf ?

**Comment by slemeur** [comment]
I'm working to get a better tray icon ;) 
Capture d’écran 2022-03-25 à 10 39 43

**Comment by evidolob** [comment]
I just copy icons from tray-electron, definitely  we need better icons

**Pull request merged**